schema-registry/serdes/src/main/java/com/hortonworks/registries/schemaregistry/serdes/avro/kafka/KafkaAvroDeserializer.java

```diff
@@ -75,33 +79,66 @@
      */
     public static final String READER_VERSIONS = "schemaregistry.reader.schema.versions";
 
-    private final AvroSnapshotDeserializer avroSnapshotDeserializer;
+    private boolean isKey;
     private Map<String, Integer> readerVersions;
 
+    private final AvroSnapshotDeserializer avroSnapshotDeserializer;
+    private final MessageContextBasedAvroDeserializer messageContextBasedAvroDeserializer;
+    private String keySchemaHeaderName;
+    private String valueSchemaHeaderName;
+
     public KafkaAvroDeserializer() {
         avroSnapshotDeserializer = new AvroSnapshotDeserializer();
+        messageContextBasedAvroDeserializer = new MessageContextBasedAvroDeserializer();
     }
 
     public KafkaAvroDeserializer(ISchemaRegistryClient schemaRegistryClient) {
         avroSnapshotDeserializer = new AvroSnapshotDeserializer(schemaRegistryClient);
+        messageContextBasedAvroDeserializer = new MessageContextBasedAvroDeserializer(schemaRegistryClient);
     }
 
     @Override
+    @SuppressWarnings("unchecked")
     public void configure(Map<String, ?> configs, boolean isKey) {
-        avroSnapshotDeserializer.init(configs);
+        this.isKey = isKey;
+        this.keySchemaHeaderName = getOrDefault(configs, KafkaAvroSerde.KEY_SCHEMA_HEADER_NAME, KafkaAvroSerde.KEY_SCHEMA_HEADER_NAME);
+        this.valueSchemaHeaderName = getOrDefault(configs, KafkaAvroSerde.VALUE_SCHEMA_HEADER_NAME, KafkaAvroSerde.VALUE_SCHEMA_HEADER_NAME);
         Map<String, Integer> versions = (Map<String, Integer>) ((Map<String, Object>) configs).get(READER_VERSIONS);
         readerVersions = versions != null ? versions : Collections.emptyMap();
+
+        avroSnapshotDeserializer.init(configs);
+        messageContextBasedAvroDeserializer.init(configs);
+    }
+
+    private static String getOrDefault(Map<String, ?> configs, String key, String defaultValue) {
```

Comment:
You may want to pull out this method as util method because it is used in `KafkaAvroSerializer` too.

Reply:
Fixed.